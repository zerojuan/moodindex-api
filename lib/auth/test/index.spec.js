import Auth, { generatePolicy, decodeToken } from '../index';
import jwt from 'jsonwebtoken';

describe( 'Authenticator', () => {
  describe( 'Policy Generation', () => {
    it( '- should return correct policy', () => {
      const fixture = {
        userId: 'user-id-1010',
        permissions: [ {
          effect: "allow",
          route: "GET/*"
        } ],
        methodArn: 'arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/<method>/<resourcePath>'
      };

      const policy = generatePolicy( fixture.userId, fixture.methodArn, fixture.permissions );

      expect( policy.principalId ).toEqual( fixture.userId );
      expect( policy.policyDocument.Statement[ 0 ].Action ).toEqual( 'execute-api:Invoke' );
      expect( policy.policyDocument.Statement[ 0 ].Effect ).toEqual( fixture.permissions[ 0 ].effect );
      expect( policy.policyDocument.Statement[ 0 ].Resource ).toEqual(
        `arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/${fixture.permissions[ 0 ].route}`
      );
      expect( policy.context.message ).toEqual( 'Authenticate Success' );
    } );
  } );

  describe( 'Token Decoding', () => {
    it( '- should decode a token using a secret', () => {
      // load sample token
      const fixture = {
        userId: 'test-id',
        roles: [ 'user:*' ]
      };

      const decodedToken = decodeToken(
        jwt.sign( fixture, process.env.JWT_SECRET )
      );

      expect( decodedToken.userId ).toBe( fixture.userId );
      expect( decodedToken.type ).toBe( fixture.type );

    } );
  } );

  describe( 'Authentication', () => {
    it( '- should return a policy response', async () => {
      const tokenFixture = {
        sub: '1',
        roles: [ "admin:*" ]
      };

      const authResult = await Auth( {
        token: jwt.sign( tokenFixture, process.env.JWT_SECRET ),
        methodArn: 'arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/<method>/<resourcePath>'
      } );

      expect( authResult ).toMatchSnapshot();
    } );
  } );
} );
