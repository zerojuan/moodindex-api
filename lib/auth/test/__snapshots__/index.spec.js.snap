// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticator Authentication - should return a policy response 1`] = `
Object {
  "context": Object {
    "message": "Authenticate Success",
  },
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "allow",
        "Resource": "arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/POST/*",
      },
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "allow",
        "Resource": "arn:aws:execute-api:<regionId>:<accountId>:<apiId>/<stage>/GET/*",
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "1",
}
`;
